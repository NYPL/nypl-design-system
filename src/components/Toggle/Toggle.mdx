import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as ToggleStories from "./Toggle.stories";
import Link from "../Link/Link";

<Meta of={ToggleStories} />

# Toggle

| Component Version | DS Version   |
| ----------------- | ------------ |
| Added             | `0.25.8`     |
| Latest            | `Prerelease` |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#sizes" target="_self">Sizes</Link>}
- {<Link href="#controlled-toggle" target="_self">Controlled Toggle</Link>}
- {<Link href="#browser-states" target="_self">Browser States</Link>}
- {<Link href="#layouts" target="_self">Layouts</Link>}

## Overview

<Description of={ToggleStories} />

The `Toggle` component is used as an alternative for the Checkbox component and
returns a boolean response. Like the Checkbox component, the `Toggle` component
can display a label to the right of the toggle and helper/error text below the
toggle.

### Best Practices

`Toggle` components are commonly used for “on/off” switches to perform binary
actions that occur immediately after a user “flips the switch.”

The Toggle component label should clarify the action being performed. Labels
should be short and to the point, often three words or less.

## Component Props

<Canvas of={ToggleStories.WithControls} />

<ArgTypes of={ToggleStories.WithControls} />

## Accessibility

The `Toggle` component is implemented using Chakra's `Switch` component. While
some accessibility resources state that using a `<button>` element is better for
a toggle or switch component, Chakra UI internally uses a `checkbox` element.
Because Chakra's `Switch` component handles the internal state so well and its
styles, we chose this over implementing this component with a `<button>` element.

For comparison and reference, resources that build toggle components with the
`<button>` element are linked below.

Resources:

- [Chakra UI Switch](https://chakra-ui.com/docs/components/form/switch)
- [W3C Switch Example Using HTML Button](https://w3c.github.io/aria-practices/examples/switch/switch-button.html)
- [Inclusive Components Toggle Buttons](https://inclusive-components.design/toggle-button/)
- [Accessible HTML Toggle Button (Switch)](https://www.erikkroes.nl/blog/accessible-html-toggle-button-switch/)

## Sizes

<Canvas of={ToggleStories.Sizes} />

## Controlled Toggle

Note that the `isChecked` property in this example is set to `true`. Clicking
on it won't change the toggle; The `isChecked` prop must now be controlled
and removed manually by the parent component that controls this state. The
`onChange` function will be called every time the `isChecked` prop changes.

<Source
  code={`
const isChecked = true;
const onChange = (e) => {
  // This will return the value through the event object.
  console.log(e.target.value);
};

<Toggle
  id="toggle"
  isChecked={true}
  labelText="Controlled Toggle"
  onChange={onChange}
/>
`} language="jsx" />

<Canvas of={ToggleStories.ControlledToggle} />

## Browser States

<Canvas of={ToggleStories.BrowserStates} />

## Layouts

<Canvas of={ToggleStories.Layouts} />
