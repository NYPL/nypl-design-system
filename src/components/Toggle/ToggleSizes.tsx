export enum ToggleSizes {
  Large = "lg",
  Small = "sm",
}
