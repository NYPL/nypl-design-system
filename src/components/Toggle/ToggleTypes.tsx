export enum ToggleSizes {
  Small = "small",
  Large = "large",
}
