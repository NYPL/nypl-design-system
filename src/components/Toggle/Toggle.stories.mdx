import { HStack, VStack, Grid, Box } from "@chakra-ui/react";
import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Description,
} from "@storybook/addon-docs";
import { withDesign } from "storybook-addon-designs";

import Heading from "../Heading/Heading";
import SimpleGrid from "../Grid/SimpleGrid";
import Toggle from "./Toggle";
import { getCategory } from "../../utils/componentCategories";
import DSProvider from "../../theme/provider";

<Meta
  title={getCategory("Toggle")}
  component={Toggle}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=31581%3A302",
    },
    jest: ["Toggle.test.tsx"],
  }}
  argTypes={{
    defaultChecked: { control: false },
    id: { control: false },
    key: { table: { disable: true } },
    name: { control: false },
    onChange: { control: false },
    ref: { table: { disable: true } },
    size: {
      control: {
        table: { defaultValue: { summary: "default" } },
      },
    },
  }}
/>

# Toggle

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.25.8`   |
| Latest            | `1.3.0`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Sizes](#sizes)
- [Controlled Toggle](#controlled-toggle)
- [Browser States](#browser-states)
- [Layouts](#layouts)

## Overview

<Description of={Toggle} />

The `Toggle` component is used as an alternative for the Checkbox component and
returns a boolean response. Like the Checkbox component, the `Toggle` component
can display a label to the right of the toggle and helper/error text below the toggle.

### Best Practices

`Toggle` components are commonly used for “on/off” switches to perform binary
actions that occur immediately after a user “flips the switch.”

The Toggle component label should clarify the action being performed. Labels
should be short and to the point, often three words or less.

## Component Props

<Canvas>
  <Story
    name="Toggle"
    args={{
      defaultChecked: false,
      helperText: "This is the helper text!",
      id: "toggle-1",
      invalidText: "",
      isChecked: undefined,
      isDisabled: false,
      isInvalid: false,
      isRequired: false,
      labelText: "Test Label",
      name: "toggle-1",
      onChange: undefined,
      size: "ToogleSizes.Large",
    }}
  >
    {(args) => <Toggle {...args} />}
  </Story>
</Canvas>

<ArgsTable story="Toggle" />

## Accessibility

The `Toggle` component is implemented using Chakra's `Switch` component. While
some accessibility resources state that using a `<button>` element is better for
a toggle or switch component, Chakra UI internally uses a `checkbox` element.
Because Chakra's `Switch` component handles the internal state so well and its
styles, we chose this over implementing this component with a `<button>` element.

For comparison and reference, resources that build toggle components with the
`<button>` element are linked below.

Resources:

- [Chakra UI Switch](https://chakra-ui.com/docs/components/form/switch)
- [W3C Switch Example Using HTML Button](https://w3c.github.io/aria-practices/examples/switch/switch-button.html)
- [Inclusive Components Toggle Buttons](https://inclusive-components.design/toggle-button/)
- [Accessible HTML Toggle Button (Switch)](https://www.erikkroes.nl/blog/accessible-html-toggle-button-switch/)

## Sizes

<Canvas>
  <DSProvider>
    <SimpleGrid columns="2">
      <VStack align="left" spacing="s">
        <Heading level={3}>Default</Heading>
        <Toggle defaultChecked={true} id="default-checked" labelText="On" />
        <Toggle id="default" labelText="Off" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level={3}>Small</Heading>
        <Toggle
          defaultChecked={true}
          id="small-checked"
          size="small"
          labelText="On"
        />
        <Toggle id="default-small" size="small" labelText="Off" />
      </VStack>
    </SimpleGrid>
  </DSProvider>
</Canvas>

## Controlled Toggle

Note that the `isChecked` property in this example is set to `true`. Clicking
on it won't change the toggle; The `isChecked` prop must now be controlled
and removed manually by the parent component that controls this state. The
`onChange` function will be called every time the `isChecked` prop changes.

```jsx
const isChecked = true;
const onChange = (e) => {
  // This will return the value through the event object.
  console.log(e.target.value);
};

<Toggle
  id="toggle"
  isChecked={true}
  labelText="Controlled Toggle"
  onChange={onChange}
/>;
```

<Canvas>
  <DSProvider>
    <Toggle
      id="toggle"
      isChecked={true}
      labelText="Controlled Toggle"
      onChange={(e) => {
        console.log(e.target.value);
      }}
    />
  </DSProvider>
</Canvas>

## Browser States

<Canvas>
  <DSProvider>
    <SimpleGrid columns="3">
      <VStack align="left" spacing="s">
        <Heading level={3}>Default</Heading>
        <Toggle defaultChecked={true} id="checked-default2" labelText="On" />
        <Toggle id="default2" labelText="Off" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level={3}>Disabled</Heading>
        <Toggle
          defaultChecked={true}
          id="disabled-checked"
          isDisabled={true}
          labelText="On"
        />
        <Toggle id="disabled2" isDisabled={true} labelText="Off" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level={3}>Error</Heading>
        <Toggle
          defaultChecked={true}
          id="error-default"
          isInvalid={true}
          labelText="On"
        />
        <Toggle id="error2" isInvalid={true} labelText="Off" />
      </VStack>
    </SimpleGrid>
  </DSProvider>
</Canvas>

## Layouts

<Canvas>
  <DSProvider>
    <SimpleGrid columns="2">
      <VStack align="left" spacing="s">
        <Heading level={3}>Grouped</Heading>
        <Toggle defaultChecked={true} id="layout1" labelText="On" />
        <Toggle defaultChecked={true} id="layout2" labelText="On" />
        <Toggle defaultChecked={true} id="layout3" labelText="On" />
        <Toggle defaultChecked={true} id="layout4" labelText="On" />
        <Toggle defaultChecked={true} id="layout5" labelText="On" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level={3}>With Helper Text</Heading>
        <Toggle
          defaultChecked={true}
          helperText="Component Helper Text"
          id="helper-text"
          labelText="On"
        />
      </VStack>
    </SimpleGrid>
  </DSProvider>
</Canvas>
