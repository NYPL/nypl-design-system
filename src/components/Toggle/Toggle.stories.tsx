import { VStack } from "@chakra-ui/react";
import type { Meta, StoryObj } from "@storybook/react";
import { withDesign } from "storybook-addon-designs";

import Heading from "../Heading/Heading";
import SimpleGrid from "../Grid/SimpleGrid";
import Toggle, { toggleSizesArray } from "./Toggle";

const meta: Meta<typeof Toggle> = {
  title: "Components/Form Elements/Toggle",
  component: Toggle,
  decorators: [withDesign],
  argTypes: {
    defaultChecked: { control: false },
    id: { control: false },
    key: { table: { disable: true } },
    name: { control: false },
    onChange: { control: false },
    ref: { table: { disable: true } },
    size: {
      control: { type: "radio" },
      options: toggleSizesArray,
      table: { defaultValue: { summary: "default" } },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Toggle>;

/**
 * Main Story for the Toggle component. This must contains the `args`
 * and `parameters` properties in this object.
 */
export const WithControls: Story = {
  args: {
    defaultChecked: false,
    helperText: "This is the helper text!",
    id: "toggle-1",
    invalidText: "",
    isChecked: undefined,
    isDisabled: false,
    isInvalid: false,
    isRequired: false,
    labelText: "Test Label",
    name: "toggle-1",
    onChange: undefined,
    size: "default",
  },
  render: (args) => <Toggle {...args} />,
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=31581%3A302",
    },
    jest: ["Toggle.test.tsx"],
  },
};

// The following are additional Toggle example Stories.
export const Sizes: Story = {
  render: () => (
    <SimpleGrid columns={2}>
      <VStack align="left" spacing="s">
        <Heading level="h3" size="heading6">
          Default
        </Heading>
        <Toggle defaultChecked={true} id="default-checked" labelText="On" />
        <Toggle id="default" labelText="Off" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level="h3" size="heading6">
          Small
        </Heading>
        <Toggle
          defaultChecked={true}
          id="small-checked"
          size="small"
          labelText="On"
        />
        <Toggle id="default-small" size="small" labelText="Off" />
      </VStack>
    </SimpleGrid>
  ),
};
export const ControlledToggle: Story = {
  render: () => (
    <Toggle
      id="toggle"
      isChecked={true}
      labelText="Controlled Toggle"
      onChange={(e) => {
        console.log(e.target.value);
      }}
    />
  ),
};

export const BrowserStates: Story = {
  render: () => (
    <SimpleGrid columns={3}>
      <VStack align="left" spacing="s">
        <Heading level="h3" size="heading6">
          Default
        </Heading>
        <Toggle defaultChecked={true} id="checked-default2" labelText="On" />
        <Toggle id="default2" labelText="Off" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level="h3" size="heading6">
          Disabled
        </Heading>
        <Toggle
          defaultChecked={true}
          id="disabled-checked"
          isDisabled={true}
          labelText="On"
        />
        <Toggle id="disabled2" isDisabled={true} labelText="Off" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level="h3" size="heading6">
          Error
        </Heading>
        <Toggle
          defaultChecked={true}
          id="error-default"
          isInvalid={true}
          labelText="On"
        />
        <Toggle id="error2" isInvalid={true} labelText="Off" />
      </VStack>
    </SimpleGrid>
  ),
};

export const Layouts: Story = {
  render: () => (
    <SimpleGrid columns={2}>
      <VStack align="left" spacing="s">
        <Heading level="h3" size="heading6">
          Grouped
        </Heading>
        <Toggle defaultChecked={true} id="layout1" labelText="On" />
        <Toggle defaultChecked={true} id="layout2" labelText="On" />
        <Toggle defaultChecked={true} id="layout3" labelText="On" />
        <Toggle defaultChecked={true} id="layout4" labelText="On" />
        <Toggle defaultChecked={true} id="layout5" labelText="On" />
      </VStack>
      <VStack align="left" spacing="s">
        <Heading level="h3" size="heading6">
          With Helper Text
        </Heading>
        <Toggle
          defaultChecked={true}
          helperText="Component Helper Text"
          id="helper-text"
          labelText="On"
        />
      </VStack>
    </SimpleGrid>
  ),
};
