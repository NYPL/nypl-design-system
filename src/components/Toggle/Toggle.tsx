import {
  Box,
  chakra,
  Switch,
  useMultiStyleConfig,
  useStyleConfig,
} from "@chakra-ui/react";
import * as React from "react";

import HelperErrorText, {
  HelperErrorTextType,
} from "../HelperErrorText/HelperErrorText";
import { ToggleSizes } from "./ToggleTypes";
import { AriaAttributes } from "../../utils/interfaces";
export interface ToggleProps {
  /** Optionally pass in additional Chakra-based styles. */
  additionalStyles?: { [key: string]: any };
  /** Used for uncontrolled scenarios.  Sets the state of the Toggle when the page first loads.
   *   If true, the toggle will be initially set to the "on" position. */
  defaultChecked?: boolean;
  /** Optional string to populate the HelperErrorText for standard state */
  helperText?: HelperErrorTextType;
  /** ID that other components can cross reference for accessibility purposes */
  id: string;
  /** Optional string to populate the HelperErrorText for the error state
   * when `isInvalid` is true. */
  invalidText?: HelperErrorTextType;
  /** When using the Toggle as a "controlled" form element, you can specify
   * the Toggle's checked state using this prop.
   * Learn more about controlled and uncontrolled form fields:
   * https://goshakkk.name/controlled-vs-uncontrolled-inputs-react/ */
  isChecked?: boolean;
  /** Adds the 'disabled' and `aria-disabled` attributes to the input when true.
   * This also makes the text italic and color scheme gray. */
  isDisabled?: boolean;
  /** Adds the 'aria-invalid' attribute to the input when true. */
  isInvalid?: boolean;
  /** Adds the 'required' attribute to the input when true. */
  isRequired?: boolean;
  /** The toggle's label. This will serve as the text content for the `<label>` element */
  labelText: string;
  /** The name prop indicates the `Toggle`'s form element name. If none is
   * specified, 'default' will be used. */
  name?: string;
  /** The action to perform on the `<input>`'s onChange function  */
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /** The size of the Toggle. Defaults to "large". */
  size?: ToggleSizes;
}

export const onChangeDefault = () => {
  return;
};

/**
 * Component that renders Chakra's `Switch` component along with NYPL defaults.
 */
export const Toggle = chakra(
  React.forwardRef<HTMLInputElement, ToggleProps>((props, ref?) => {
    const {
      additionalStyles = {},
      defaultChecked = false,
      helperText,
      id,
      invalidText,
      isChecked,
      isDisabled = false,
      isInvalid = false,
      isRequired = false,
      labelText,
      name,
      onChange = onChangeDefault,
      size = ToggleSizes.Default,
      ...rest
    } = props;
    const footnote = isInvalid ? invalidText : helperText;
    const ariaAttributes: AriaAttributes = {};
    const styles = useMultiStyleConfig("Toggle", { isDisabled, size });
    const switchStyles = useStyleConfig("Switch", { size });
    ariaAttributes["aria-label"] =
      labelText && footnote ? `${labelText} - ${footnote}` : labelText;

    if (!id) {
      console.warn(
        "NYPL Reservoir Toggle: This component's required `id` prop was not passed."
      );
    }

    return (
      <>
        <Box __css={{ ...styles, ...additionalStyles }} {...rest}>
          <Switch
            id={id}
            name={name || "default"}
            isDisabled={isDisabled}
            isInvalid={isInvalid}
            isRequired={isRequired}
            ref={ref}
            size={size === ToggleSizes.Default ? "lg" : "sm"}
            lineHeight="1.5"
            {...(isChecked !== undefined
              ? {
                  isChecked,
                  onChange,
                }
              : {
                  defaultChecked,
                })}
            {...ariaAttributes}
            __css={switchStyles}
          >
            {labelText}
          </Switch>
        </Box>
        {footnote && (
          <HelperErrorText
            additionalStyles={styles.helperErrorText}
            id={`${id}-helperText`}
            isInvalid={isInvalid}
            text={footnote}
          />
        )}
      </>
    );
  })
);

export default Toggle;
