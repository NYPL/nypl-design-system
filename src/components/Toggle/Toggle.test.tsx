import * as React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe } from "jest-axe";
import renderer from "react-test-renderer";

import * as generateUUID from "../../helpers/generateUUID";
import Toggle from "./Toggle";

describe("Toggle Accessibility", () => {
  it("Passes axe accessibility test", async () => {
    const { container } = render(
      <Toggle id="inputID" onChange={jest.fn()} labelText="Test Label" />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("Toggle", () => {
  let changeHandler;
  let generateUUIDSpy;

  beforeEach(() => {
    changeHandler = jest.fn();
  });

  it("Renders with a Toggle input and label", () => {
    render(<Toggle id="inputID" labelText="Test Label" />);
    expect(screen.getByLabelText("Test Label")).toBeInTheDocument();
    expect(screen.getByRole("checkbox")).toBeInTheDocument();
  });

  it("Renders visible helper or error text", () => {
    const { rerender } = render(
      <Toggle
        id="inputID"
        labelText="Test Label"
        helperText="This is the helper text."
        invalidText="This is the error text :("
      />
    );
    expect(screen.getByText("This is the helper text.")).toBeVisible();
    expect(
      screen.queryByText("This is the error text :(")
    ).not.toBeInTheDocument();

    rerender(
      <Toggle
        id="inputID"
        labelText="Test Label"
        isInvalid
        helperText="This is the helper text."
        invalidText="This is the error text :("
      />
    );
    expect(screen.getByText("This is the error text :(")).toBeVisible();
    expect(
      screen.queryByText("This is the helper text.")
    ).not.toBeInTheDocument();
  });

  it("Sets the Toggle's ID", () => {
    render(<Toggle id="inputID" labelText="Test Label" />);
    expect(screen.getByRole("checkbox")).toHaveAttribute("id", "inputID");
  });

  it("Calls the UUID generation function if no id prop value is passed", () => {
    generateUUIDSpy = jest.spyOn(generateUUID, "default");
    expect(generateUUIDSpy).toHaveBeenCalledTimes(0);
    render(<Toggle labelText="Test Label" />);
    expect(generateUUIDSpy).toHaveBeenCalledTimes(1);
  });

  it("Sets the 'checked' attribute", () => {
    render(<Toggle id="inputID" labelText="Test Label" isChecked />);
    expect(screen.getByRole("checkbox")).toHaveAttribute("checked");
  });

  it("Sets the 'disabled' attribute", () => {
    render(<Toggle id="inputID" labelText="Test Label" isDisabled />);
    expect(screen.getByRole("checkbox")).toHaveAttribute("disabled");
  });

  it("Sets the error state", () => {
    render(
      <Toggle
        id="inputID-attributes"
        labelText="onChange test"
        invalidText="This is the error text!"
        isInvalid
      />
    );
    expect(screen.getByRole("checkbox")).toHaveAttribute("aria-invalid");
    expect(screen.getByText("This is the error text!")).toBeInTheDocument();
  });

  it("does not render the error text when 'isInvalid' is true but 'invalidText' is not passed", () => {
    render(
      <Toggle id="inputID-attributes" labelText="onChange test" isInvalid />
    );
    expect(screen.getByRole("checkbox")).toHaveAttribute("aria-invalid");
    expect(
      screen.queryByText("This is the error text!")
    ).not.toBeInTheDocument();
  });

  it("Changing the value calls the onChange handler", () => {
    const utils = render(
      <Toggle
        id="onChangeTest"
        onChange={changeHandler}
        labelText="onChangeTest Lab"
        isChecked
      />
    );

    expect(changeHandler).toHaveBeenCalledTimes(0);
    userEvent.click(utils.getByText("onChangeTest Lab"));
    expect(changeHandler).toHaveBeenCalledTimes(1);
  });

  it("Renders the UI snapshot correctly", () => {
    const primary = renderer
      .create(<Toggle id="inputID" labelText="Test Label" />)
      .toJSON();
    const isChecked = renderer
      .create(<Toggle id="Toggle-checked" labelText="Test Label" isChecked />)
      .toJSON();
    const isInvalid = renderer
      .create(<Toggle id="Toggle-invalid" labelText="Test Label" isInvalid />)
      .toJSON();
    const isDisabled = renderer
      .create(<Toggle id="Toggle-disabled" labelText="Test Label" isDisabled />)
      .toJSON();

    expect(primary).toMatchSnapshot();
    expect(isChecked).toMatchSnapshot();
    expect(isInvalid).toMatchSnapshot();
    expect(isDisabled).toMatchSnapshot();
  });
});
