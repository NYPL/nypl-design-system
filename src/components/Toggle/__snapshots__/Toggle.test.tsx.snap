// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Renders the UI snapshot correctly 1`] = `
<div
  className="css-1u8qly9"
  id="inputID-wrapper"
>
  <div
    className="css-0"
  >
    <label
      className="chakra-switch css-18x0has"
      onClick={[Function]}
    >
      <input
        aria-disabled={false}
        aria-invalid={false}
        checked={false}
        className="chakra-switch__input"
        disabled={false}
        id="inputID"
        name="default"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        style={
          {
            "border": "0px",
            "clip": "rect(0px, 0px, 0px, 0px)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": "0px",
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "1px",
          }
        }
        type="checkbox"
      />
      <span
        aria-hidden={true}
        className="chakra-switch__track css-14qxnv8"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
      >
        <span
          className="chakra-switch__thumb css-0"
        />
      </span>
      <span
        className="chakra-switch__label css-1y8kf23"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        Test Label
      </span>
    </label>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="inputID-helperText"
  />
</div>
`;

exports[`Toggle Renders the UI snapshot correctly 2`] = `
<div
  className="css-1u8qly9"
  id="Toggle-checked-wrapper"
>
  <div
    className="css-0"
  >
    <label
      className="chakra-switch css-18x0has"
      data-checked=""
      onClick={[Function]}
    >
      <input
        aria-disabled={false}
        aria-invalid={false}
        checked={true}
        className="chakra-switch__input"
        disabled={false}
        id="Toggle-checked"
        name="default"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        style={
          {
            "border": "0px",
            "clip": "rect(0px, 0px, 0px, 0px)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": "0px",
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "1px",
          }
        }
        type="checkbox"
      />
      <span
        aria-hidden={true}
        className="chakra-switch__track css-14qxnv8"
        data-checked=""
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
      >
        <span
          className="chakra-switch__thumb css-0"
          data-checked=""
        />
      </span>
      <span
        className="chakra-switch__label css-1y8kf23"
        data-checked=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        Test Label
      </span>
    </label>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="Toggle-checked-helperText"
  />
</div>
`;

exports[`Toggle Renders the UI snapshot correctly 3`] = `
<div
  className="css-1u8qly9"
  id="Toggle-invalid-wrapper"
>
  <div
    className="css-0"
  >
    <label
      className="chakra-switch css-18x0has"
      data-invalid=""
      onClick={[Function]}
    >
      <input
        aria-disabled={false}
        aria-invalid={true}
        checked={false}
        className="chakra-switch__input"
        disabled={false}
        id="Toggle-invalid"
        name="default"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        style={
          {
            "border": "0px",
            "clip": "rect(0px, 0px, 0px, 0px)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": "0px",
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "1px",
          }
        }
        type="checkbox"
      />
      <span
        aria-hidden={true}
        className="chakra-switch__track css-14qxnv8"
        data-invalid=""
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
      >
        <span
          className="chakra-switch__thumb css-0"
        />
      </span>
      <span
        className="chakra-switch__label css-1y8kf23"
        data-invalid=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        Test Label
      </span>
    </label>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={true}
    id="Toggle-invalid-helperText"
  />
</div>
`;

exports[`Toggle Renders the UI snapshot correctly 4`] = `
<div
  className="css-1u8qly9"
  id="Toggle-disabled-wrapper"
>
  <div
    className="css-0"
  >
    <label
      className="chakra-switch css-18x0has"
      data-disabled=""
      onClick={[Function]}
    >
      <input
        aria-disabled={true}
        aria-invalid={false}
        checked={false}
        className="chakra-switch__input"
        disabled={true}
        id="Toggle-disabled"
        name="default"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        style={
          {
            "border": "0px",
            "clip": "rect(0px, 0px, 0px, 0px)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": "0px",
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "1px",
          }
        }
        type="checkbox"
      />
      <span
        aria-hidden={true}
        className="chakra-switch__track css-14qxnv8"
        data-disabled=""
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
      >
        <span
          className="chakra-switch__thumb css-0"
        />
      </span>
      <span
        className="chakra-switch__label css-1y8kf23"
        data-disabled=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        Test Label
      </span>
    </label>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="Toggle-disabled-helperText"
  />
</div>
`;

exports[`Toggle Renders the UI snapshot correctly 5`] = `
<div
  className="css-1u8qly9"
  id="Toggle-disabled-wrapper"
>
  <div
    className="css-0"
  >
    <label
      className="chakra-switch css-18x0has"
      data-disabled=""
      onClick={[Function]}
    >
      <input
        aria-disabled={true}
        aria-invalid={false}
        checked={false}
        className="chakra-switch__input"
        disabled={true}
        id="Toggle-disabled"
        name="default"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        style={
          {
            "border": "0px",
            "clip": "rect(0px, 0px, 0px, 0px)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": "0px",
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "1px",
          }
        }
        type="checkbox"
      />
      <span
        aria-hidden={true}
        className="chakra-switch__track css-14qxnv8"
        data-disabled=""
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
      >
        <span
          className="chakra-switch__thumb css-0"
        />
      </span>
      <span
        className="chakra-switch__label css-1y8kf23"
        data-disabled=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        Test Label
      </span>
    </label>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="Toggle-disabled-helperText"
  />
</div>
`;

exports[`Toggle Renders the UI snapshot correctly 6`] = `
<div
  className="css-7zjd3y"
  id="chakra-wrapper"
>
  <div
    className="css-0"
  >
    <label
      className="chakra-switch css-18x0has"
      onClick={[Function]}
    >
      <input
        aria-disabled={false}
        aria-invalid={false}
        checked={false}
        className="chakra-switch__input"
        disabled={false}
        id="chakra"
        name="default"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        style={
          {
            "border": "0px",
            "clip": "rect(0px, 0px, 0px, 0px)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": "0px",
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "1px",
          }
        }
        type="checkbox"
      />
      <span
        aria-hidden={true}
        className="chakra-switch__track css-14qxnv8"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
      >
        <span
          className="chakra-switch__thumb css-0"
        />
      </span>
      <span
        className="chakra-switch__label css-1y8kf23"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        Test Label
      </span>
    </label>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="chakra-helperText"
  />
</div>
`;

exports[`Toggle Renders the UI snapshot correctly 7`] = `
<div
  className="css-1u8qly9"
  data-testid="props"
  id="props-wrapper"
>
  <div
    className="css-0"
  >
    <label
      className="chakra-switch css-18x0has"
      onClick={[Function]}
    >
      <input
        aria-disabled={false}
        aria-invalid={false}
        checked={false}
        className="chakra-switch__input"
        disabled={false}
        id="props"
        name="default"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        style={
          {
            "border": "0px",
            "clip": "rect(0px, 0px, 0px, 0px)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": "0px",
            "position": "absolute",
            "whiteSpace": "nowrap",
            "width": "1px",
          }
        }
        type="checkbox"
      />
      <span
        aria-hidden={true}
        className="chakra-switch__track css-14qxnv8"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
      >
        <span
          className="chakra-switch__thumb css-0"
        />
      </span>
      <span
        className="chakra-switch__label css-1y8kf23"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        Test Label
      </span>
    </label>
  </div>
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    data-isinvalid={false}
    id="props-helperText"
  />
</div>
`;
